src/lib/operators/join_hash/multi_predicate_join_evaluator.hpp

```diff
@@ -0,0 +1,145 @@
+#include "types.hpp"
+
+namespace opossum {
+
+class BaseFieldComparator : public Noncopyable {
+ public:
+  virtual bool compare(const RowID& left, const RowID& right) const = 0;
+  virtual ~BaseFieldComparator() = default;
+};
+
+template <typename CompareFunctor, typename L, typename R>
+class FieldComparator : public BaseFieldComparator {
+ public:
+  FieldComparator(CompareFunctor compare_functor,
+                  std::vector<std::unique_ptr<AbstractSegmentAccessor<L>>> left_accessors,
+                  std::vector<std::unique_ptr<AbstractSegmentAccessor<R>>> right_accessors)
+      : _compare{std::move(compare_functor)},
+        _left_accessors{std::move(left_accessors)},
+        _right_accessors{std::move(right_accessors)} {}
+
+  bool compare(const RowID& left, const RowID& right) const {
+    // Todo: Expensive copy operations?
+    const auto left_opt = _left_accessors[left.chunk_id]->access(left.chunk_offset);
+    const auto right_opt = _right_accessors[right.chunk_id]->access(right.chunk_offset);
+
+    // Todo: What, if we don't have a value here?
+    return _compare(left_opt.value(), right_opt.value());
+  }
+
+ private:
+  const CompareFunctor _compare;
+  const std::vector<std::unique_ptr<AbstractSegmentAccessor<L>>> _left_accessors;
+  const std::vector<std::unique_ptr<AbstractSegmentAccessor<R>>> _right_accessors;
+};
+
+class MultiPredicateJoinEvaluator {
+ public:
+  MultiPredicateJoinEvaluator(const Table& left, const Table& right,
+                              const std::vector<JoinPredicate>& join_predicates) {
+    for (const auto& pred : join_predicates) {
+      resolve_data_type(left.column_data_type(pred.column_id_pair.first), [&](auto left_type) {
+        using LeftColumnDataType = typename decltype(left_type)::type;
+        resolve_data_type(right.column_data_type(pred.column_id_pair.second), [&](auto right_type) {
+          using RightColumnDataType = typename decltype(right_type)::type;
+
+          std::vector<std::unique_ptr<AbstractSegmentAccessor<LeftColumnDataType>>> left_accessors;
+          std::vector<std::unique_ptr<AbstractSegmentAccessor<RightColumnDataType>>> right_accessors;
+
+          left_accessors = _create_accessors<LeftColumnDataType>(left, pred.column_id_pair.first);
+          right_accessors = _create_accessors<RightColumnDataType>(right, pred.column_id_pair.second);
+
+          // todo: copied from column_vs_column_table_scan_impl
+          constexpr auto LEFT_IS_STRING_COLUMN = (std::is_same<LeftColumnDataType, std::string>{});
+          constexpr auto RIGHT_IS_STRING_COLUMN = (std::is_same<RightColumnDataType, std::string>{});
+
+          constexpr auto NEITHER_IS_STRING_COLUMN = !LEFT_IS_STRING_COLUMN && !RIGHT_IS_STRING_COLUMN;
+          constexpr auto BOTH_ARE_STRING_COLUMNS = LEFT_IS_STRING_COLUMN && RIGHT_IS_STRING_COLUMN;
+
+          if constexpr (NEITHER_IS_STRING_COLUMN || BOTH_ARE_STRING_COLUMNS) {
+            switch (pred.predicate_condition) {
```

Comment:
with_comparator(_light)?

Reply:
Honestly, I don't know how to use with_comparator in our case. I tried the following:
```
MultiPredicateJoinEvaluator(const Table& left, const Table& right,
                              const std::vector<JoinPredicate>& join_predicates) {

    std::vector<std::unique_ptr<BaseFieldComparator>> comparators;

    for (const auto& predicate : join_predicates) {
      resolve_data_type(left.column_data_type(predicate.column_id_pair.first), [&](auto left_type) {
        using LeftColumnDataType = typename decltype(left_type)::type;
        resolve_data_type(right.column_data_type(predicate.column_id_pair.second), [&](auto right_type) {
          using RightColumnDataType = typename decltype(right_type)::type;

          // todo: copied from column_vs_column_table_scan_impl
          constexpr auto LEFT_IS_STRING_COLUMN = (std::is_same<LeftColumnDataType, std::string>{});
          constexpr auto RIGHT_IS_STRING_COLUMN = (std::is_same<RightColumnDataType, std::string>{});

          constexpr auto NEITHER_IS_STRING_COLUMN = !LEFT_IS_STRING_COLUMN && !RIGHT_IS_STRING_COLUMN;
          constexpr auto BOTH_ARE_STRING_COLUMNS = LEFT_IS_STRING_COLUMN && RIGHT_IS_STRING_COLUMN;

          if constexpr (NEITHER_IS_STRING_COLUMN || BOTH_ARE_STRING_COLUMNS) {
            with_comparator(predicate.predicate_condition, [&comparators, &left, &right, &predicate](auto comparator) {

              std::vector<std::unique_ptr<AbstractSegmentAccessor<LeftColumnDataType>>> left_accessors;
              std::vector<std::unique_ptr<AbstractSegmentAccessor<RightColumnDataType>>> right_accessors;

              left_accessors = _create_accessors<LeftColumnDataType>(left, predicate.column_id_pair.first);
              right_accessors = _create_accessors<RightColumnDataType>(right, predicate.column_id_pair.second);


              comparators.emplace_back(
                  std::make_unique<decltype(comparator), LeftColumnDataType, RightColumnDataType>>(
                      std::move(comparator), std::move(left_accessors), std::move(right_accessors)));
            });

          } else {
            Fail("Types of columns cannot be compared.");
          }
        });
      });
    }

    _comparators = std::move(comparators);
  }
```
This does not compile due to an internal compiler error.

```
In file included from /tmp/tmp.fPjVrYwXCJ/src/lib/operators/join_hash/join_hash_steps.hpp:7,
                 from /tmp/tmp.fPjVrYwXCJ/src/lib/operators/join_hash.cpp:13:
/tmp/tmp.fPjVrYwXCJ/src/lib/operators/multi_predicate_join_evaluator.hpp: In instantiation of ‘opossum::MultiPredicateJoinEvaluator::MultiPredicateJoinEvaluator(const opossum::Table&, const opossum::Table&, const std::vector<opossum::JoinPredicate>&)::<lambda(auto:17)> [with auto:17 = boost::hana::basic_type<int>]::<lambda(auto:18)> [with auto:18 = boost::hana::basic_type<int>]’:
/tmp/tmp.fPjVrYwXCJ/src/lib/resolve_type.hpp:165:11:   required from ‘opossum::resolve_data_type(opossum::DataType, const Functor&) [with Functor = opossum::MultiPredicateJoinEvaluator::MultiPredicateJoinEvaluator(const opossum::Table&, const opossum::Table&, const std::vector<opossum::JoinPredicate>&)::<lambda(auto:17)> [with auto:17 = boost::hana::basic_type<int>]::<lambda(auto:18)>; opossum::DataType = opossum::detail::DataType]::<lambda(auto:12)> [with auto:12 = boost::hana::pair<opossum::detail::DataType, boost::hana::type_impl<int>::_>]’
/usr/include/boost/hana/for_each.hpp:46:45:   required from ‘constexpr void boost::hana::detail::on_each<F>::operator()(Xs&& ...) const [with Xs = {const boost::hana::pair<opossum::detail::DataType, boost::hana::type_impl<int>::_>&, const boost::hana::pair<opossum::detail::DataType, boost::hana::type_impl<long int>::_>&, const boost::hana::pair<opossum::detail::DataType, boost::hana::type_impl<float>::_>&, const boost::hana::pair<opossum::detail::DataType, boost::hana::type_impl<double>::_>&, const boost::hana::pair<opossum::detail::DataType, boost::hana::type_impl<std::__cxx11::basic_string<char> >::_>&}; F = opossum::resolve_data_type(opossum::DataType, const Functor&) [with Functor = opossum::MultiPredicateJoinEvaluator::MultiPredicateJoinEvaluator(const opossum::Table&, const opossum::Table&, const std::vector<opossum::JoinPredicate>&)::<lambda(auto:17)> [with auto:17 = boost::hana::basic_type<int>]::<lambda(auto:18)>; opossum::DataType = opossum::detail::DataType]::<lambda(auto:12)>*]’
/usr/include/boost/hana/basic_tuple.hpp:107:39:   required from ‘static constexpr decltype(auto) boost::hana::unpack_impl<boost::hana::basic_tuple_tag>::apply(const boost::hana::detail::basic_tuple_impl<std::integer_sequence<long unsigned int, _Idx ...>, Xn ...>&, F&&) [with long unsigned int ...i = {0, 1, 2, 3, 4}; Xn = {boost::hana::pair<opossum::detail::DataType, boost::hana::type_impl<int>::_>, boost::hana::pair<opossum::detail::DataType, boost::hana::type_impl<long int>::_>, boost::hana::pair<opossum::detail::DataType, boost::hana::type_impl<float>::_>, boost::hana::pair<opossum::detail::DataType, boost::hana::type_impl<double>::_>, boost::hana::pair<opossum::detail::DataType, boost::hana::type_impl<std::__cxx11::basic_string<char> >::_>}; F = boost::hana::detail::on_each<opossum::resolve_data_type(opossum::DataType, const Functor&) [with Functor = opossum::MultiPredicateJoinEvaluator::MultiPredicateJoinEvaluator(const opossum::Table&, const opossum::Table&, const std::vector<opossum::JoinPredicate>&)::<lambda(auto:17)> [with auto:17 = boost::hana::basic_type<int>]::<lambda(auto:18)>; opossum::DataType = opossum::detail::DataType]::<lambda(auto:12)>*>]’
/usr/include/boost/hana/unpack.hpp:47:29:   required from ‘constexpr decltype(auto) boost::hana::unpack_t::operator()(Xs&&, F&&) const [with Xs = const boost::hana::basic_tuple<boost::hana::pair<opossum::detail::DataType, boost::hana::type_impl<int>::_>, boost::hana::pair<opossum::detail::DataType, boost::hana::type_impl<long int>::_>, boost::hana::pair<opossum::detail::DataType, boost::hana::type_impl<float>::_>, boost::hana::pair<opossum::detail::DataType, boost::hana::type_impl<double>::_>, boost::hana::pair<opossum::detail::DataType, boost::hana::type_impl<std::__cxx11::basic_string<char> >::_> >&; F = boost::hana::detail::on_each<opossum::resolve_data_type(opossum::DataType, const Functor&) [with Functor = opossum::MultiPredicateJoinEvaluator::MultiPredicateJoinEvaluator(const opossum::Table&, const opossum::Table&, const std::vector<opossum::JoinPredicate>&)::<lambda(auto:17)> [with auto:17 = boost::hana::basic_type<int>]::<lambda(auto:18)>; opossum::DataType = opossum::detail::DataType]::<lambda(auto:12)>*>]’
/usr/include/boost/hana/tuple.hpp:225:32:   required from ‘static constexpr decltype(auto) boost::hana::unpack_impl<boost::hana::tuple_tag>::apply(Xs&&, F&&) [with Xs = const boost::hana::tuple<boost::hana::pair<opossum::detail::DataType, boost::hana::type_impl<int>::_>, boost::hana::pair<opossum::detail::DataType, boost::hana::type_impl<long int>::_>, boost::hana::pair<opossum::detail::DataType, boost::hana::type_impl<float>::_>, boost::hana::pair<opossum::detail::DataType, boost::hana::type_impl<double>::_>, boost::hana::pair<opossum::detail::DataType, boost::hana::type_impl<std::__cxx11::basic_string<char> >::_> >&; F = boost::hana::detail::on_each<opossum::resolve_data_type(opossum::DataType, const Functor&) [with Functor = opossum::MultiPredicateJoinEvaluator::MultiPredicateJoinEvaluator(const opossum::Table&, const opossum::Table&, const std::vector<opossum::JoinPredicate>&)::<lambda(auto:17)> [with auto:17 = boost::hana::basic_type<int>]::<lambda(auto:18)>; opossum::DataType = opossum::detail::DataType]::<lambda(auto:12)>*>]’
/usr/include/boost/hana/unpack.hpp:47:29:   [ skipping 9 instantiation contexts, use -ftemplate-backtrace-limit=0 to disable ]
/usr/include/boost/hana/tuple.hpp:225:32:   required from ‘static constexpr decltype(auto) boost::hana::unpack_impl<boost::hana::tuple_tag>::apply(Xs&&, F&&) [with Xs = const boost::hana::tuple<boost::hana::pair<opossum::detail::DataType, boost::hana::type_impl<int>::_>, boost::hana::pair<opossum::detail::DataType, boost::hana::type_impl<long int>::_>, boost::hana::pair<opossum::detail::DataType, boost::hana::type_impl<float>::_>, boost::hana::pair<opossum::detail::DataType, boost::hana::type_impl<double>::_>, boost::hana::pair<opossum::detail::DataType, boost::hana::type_impl<std::__cxx11::basic_string<char> >::_> >&; F = boost::hana::detail::on_each<opossum::resolve_data_type(opossum::DataType, const Functor&) [with Functor = opossum::MultiPredicateJoinEvaluator::MultiPredicateJoinEvaluator(const opossum::Table&, const opossum::Table&, const std::vector<opossum::JoinPredicate>&)::<lambda(auto:17)>; opossum::DataType = opossum::detail::DataType]::<lambda(auto:12)>*>]’
/usr/include/boost/hana/unpack.hpp:47:29:   required from ‘constexpr decltype(auto) boost::hana::unpack_t::operator()(Xs&&, F&&) const [with Xs = const boost::hana::tuple<boost::hana::pair<opossum::detail::DataType, boost::hana::type_impl<int>::_>, boost::hana::pair<opossum::detail::DataType, boost::hana::type_impl<long int>::_>, boost::hana::pair<opossum::detail::DataType, boost::hana::type_impl<float>::_>, boost::hana::pair<opossum::detail::DataType, boost::hana::type_impl<double>::_>, boost::hana::pair<opossum::detail::DataType, boost::hana::type_impl<std::__cxx11::basic_string<char> >::_> >&; F = boost::hana::detail::on_each<opossum::resolve_data_type(opossum::DataType, const Functor&) [with Functor = opossum::MultiPredicateJoinEvaluator::MultiPredicateJoinEvaluator(const opossum::Table&, const opossum::Table&, const std::vector<opossum::JoinPredicate>&)::<lambda(auto:17)>; opossum::DataType = opossum::detail::DataType]::<lambda(auto:12)>*>]’
/usr/include/boost/hana/for_each.hpp:56:25:   required from ‘static constexpr void boost::hana::for_each_impl<T, boost::hana::when<condition> >::apply(Xs&&, F&&) [with Xs = const boost::hana::tuple<boost::hana::pair<opossum::detail::DataType, boost::hana::type_impl<int>::_>, boost::hana::pair<opossum::detail::DataType, boost::hana::type_impl<long int>::_>, boost::hana::pair<opossum::detail::DataType, boost::hana::type_impl<float>::_>, boost::hana::pair<opossum::detail::DataType, boost::hana::type_impl<double>::_>, boost::hana::pair<opossum::detail::DataType, boost::hana::type_impl<std::__cxx11::basic_string<char> >::_> >&; F = opossum::resolve_data_type(opossum::DataType, const Functor&) [with Functor = opossum::MultiPredicateJoinEvaluator::MultiPredicateJoinEvaluator(const opossum::Table&, const opossum::Table&, const std::vector<opossum::JoinPredicate>&)::<lambda(auto:17)>; opossum::DataType = opossum::detail::DataType]::<lambda(auto:12)>; T = boost::hana::tuple_tag; bool condition = true]’
/usr/include/boost/hana/for_each.hpp:35:30:   required from ‘constexpr void boost::hana::for_each_t::operator()(Xs&&, F&&) const [with Xs = const boost::hana::tuple<boost::hana::pair<opossum::detail::DataType, boost::hana::type_impl<int>::_>, boost::hana::pair<opossum::detail::DataType, boost::hana::type_impl<long int>::_>, boost::hana::pair<opossum::detail::DataType, boost::hana::type_impl<float>::_>, boost::hana::pair<opossum::detail::DataType, boost::hana::type_impl<double>::_>, boost::hana::pair<opossum::detail::DataType, boost::hana::type_impl<std::__cxx11::basic_string<char> >::_> >&; F = opossum::resolve_data_type(opossum::DataType, const Functor&) [with Functor = opossum::MultiPredicateJoinEvaluator::MultiPredicateJoinEvaluator(const opossum::Table&, const opossum::Table&, const std::vector<opossum::JoinPredicate>&)::<lambda(auto:17)>; opossum::DataType = opossum::detail::DataType]::<lambda(auto:12)>]’
/tmp/tmp.fPjVrYwXCJ/src/lib/resolve_type.hpp:162:17:   required from ‘void opossum::resolve_data_type(opossum::DataType, const Functor&) [with Functor = opossum::MultiPredicateJoinEvaluator::MultiPredicateJoinEvaluator(const opossum::Table&, const opossum::Table&, const std::vector<opossum::JoinPredicate>&)::<lambda(auto:17)>; opossum::DataType = opossum::detail::DataType]’
/tmp/tmp.fPjVrYwXCJ/src/lib/operators/multi_predicate_join_evaluator.hpp:85:8:   required from here
/tmp/tmp.fPjVrYwXCJ/src/lib/operators/multi_predicate_join_evaluator.hpp:67:62: internal compiler error: in tsubst_copy, at cp/pt.c:15347
             with_comparator(predicate.predicate_condition, [&comparators, &left, &right, &predicate](auto comparator) {
                                                              ^~~~~~~~~~~
Please submit a full bug report,
with preprocessed source if appropriate.
See <file:///usr/share/doc/gcc-8/README.Bugs> for instructions.
```

Reply:
This is probably https://gcc.gnu.org/bugzilla/show_bug.cgi?id=86740
You will need a copy or reference inside the lambda to avoid the compiler bug, see column_vs_column_table_scan_impl for an example. Other than that, this looks good on first sight.